scintillometry/psrio/psrfits_io.py

```diff
@@ -0,0 +1,383 @@
+        self.verify()
+        self.offset = 0
+
+    def _init_empty(self):
+        for k in self._header_defaults:
+            self.header.set(k[0], k[1], k[2])
+
+    def make_columns(self, name):
+        cols = []
+        for cl in self._columns_defaults:
+            col = fits.Column()
+
+    def verify(self):
+        if self.header['EXTNAME'].replace(' ', '') != "SUBINT":
+            raise ValueError("Input HDU is not a SUBINT type.")
+
+    @property
+    def mode(self):
+        return self.header_hdu.obs_mode
+
+    @property
+    def sample_rate(self):
+        # We assume the sample rate are the same for all the rows.
+        sample_time = u.Quantity(self.data[0]['TSUBINT'] /
+                                 self.samples_per_frame, u.s)
+        return 1.0 / sample_time
+
+    @sample_rate.setter
+    def sample_rate(self, val):
+        if hasattr(val, 'unit'):
+            val = val.to(u.Hz).value
+        self.header['TBIN'] = 1.0 / val
+
+    @property
+    def start_time(self):
+        # NOTE should we get the start time for each raw, in case the time gaps
+        # in between the rows
+        file_start = self.header_hdu.start_time
+        if "OFFS_SUB" in self.columns.names:
+            subint_times = u.Quantity(self.data['OFFS_SUB'], u.s, copy=False)
+            sample_time = 1.0 / self.sample_rate
+            start_time = (file_start + subint_times[0] -
+                          self.samples_per_frame / 2 * sample_time)
+        else:
+            start_time = file_start
+        return start_time
+
+    @start_time.setter
+    def start_time(self, time):
+        # NOTE this sets the start time of the HDU, not the file start time.
+        dt = (time - self.header_hdu.start_time).to(u.s)
+        center_off = dt + 1.0 / self.sample_rate * self.samples_per_frame / 2
+        if "OFFS_SUB" in self.columns.names:
+            first_off = self.data['OFFS_SUB'][0]
+            self.data['OFFS_SUB'] = (self.data['OFFS_SUB'] - first_off +
+                                     center_off)
+        else:
+            self._req_columns['OFFS_SUB']['value'] = center_off
+
+    @property
+    def nrow(self):
+        return int(self.header['NAXIS2'])
+
+    @property
+    def nchan(self):
+        return int(self.header['NCHAN'])
+
+    @property
+    def npol(self):
+        return int(self.header['NPOL'])
+
+    @property
+    def nbin(self):
+        return int(self.header['NBIN'])
+
+    @property
+    def samples_per_frame(self):
+        try:
+            return int(self.header['NSBLK'])
+        except:
+            return int(np.prod(self.data_shape)/(self.nrow * self.nchan *
+                                                 self.npol * self.nbin))
+
+    @samples_per_frame.setter
+    def samples_per_frame(self, val):
+        self.header['NSBLK'] = val
+
+    @property
+    def shape(self):
+        raw_shape = self.raw_shape
+        new_shape = namedtuple('shape', ['nsample','nbin', 'nchan', 'npol'])
```

Comment:
I think the names are most important for `sample_shape`, and like in `dada/payload`, I'd just define a class wide
```
    _sample_shape_maker = namedtuple('SampleShape', 'nbin, npol, nchan')
```
I don't think the regular shape has to be a `namedtuple`, though perhaps it is nice too.